arangod/VocBase/Methods/Databases.cpp

```diff
@@ -124,185 +252,175 @@ arangodb::Result Databases::info(TRI_vocbase_t* vocbase, VPackBuilder& result) {
-        return TRI_ERROR_NO_ERROR;
-      });
-    }
+  Result res = grantCurrentUser(info);
+  if (!res.ok()) {
+    return res;
+  }
 
-    TRI_ASSERT(sanitizedUsers.slice().isArray());
-    upgradeRes = methods::Upgrade::createDB(*vocbase, sanitizedUsers.slice());
-  } else {  // Single, DBServer, Agency
-    // options for database (currently only allows setting "id"
-    // for testing purposes)
-    TRI_voc_tick_t id = 0;
-    if (options.hasKey("id")) {
-      id = basics::VelocyPackHelper::stringUInt64(options, "id");
-    }
+  UpgradeResult upgradeRes = methods::Upgrade::createDB(*vocbase, info.getUsers());
 
-    TRI_vocbase_t* vocbase = nullptr;
-    int res = databaseFeature->createDatabase(id, dbName, vocbase);
-    if (res != TRI_ERROR_NO_ERROR) {
-      return Result(res);
-    }
+  return std::move(upgradeRes.result());
+}
 
-    TRI_ASSERT(vocbase != nullptr);
-    TRI_ASSERT(!vocbase->isDangling());
+arangodb::Result Databases::create(std::string const& dbName, VPackSlice const& users,
+                                   VPackSlice const& options) {
+  arangodb::Result res;
 
-    TRI_DEFER(vocbase->release());
+  // Only admin users are permitted to create databases
+  ExecContext const& exec = ExecContext::current();
+  if (!exec.isAdminUser()) {
+    events::CreateDatabase(dbName, TRI_ERROR_FORBIDDEN);
+    return Result(TRI_ERROR_FORBIDDEN);
+  }
 
-    // we need to add the permissions before running the upgrade script
-    ExecContext const& exec = ExecContext::current();
-    if (!exec.isSuperuser() && um != nullptr) {
-      // ignore errors here Result r =
-      um->updateUser(exec.user(), [&](auth::User& entry) {
-        entry.grantDatabase(dbName, auth::Level::RW);
-        entry.grantCollection(dbName, "*", auth::Level::RW);
-        return TRI_ERROR_NO_ERROR;
-      });
-    }
+  // Encapsulate and sanitize the input
+  // TODO: maybe this should just be a function that produces
+  //       a struct to avoid the try/catch
+  //       or the object could have a .valid() method?
+  CreateDatabaseInfo createInfo;
+  try {
+      createInfo = CreateDatabaseInfo(dbName, options, users);
+  } catch(Result& e) {
+    return std::move(e);
+  } catch (...) {
+    LOG_TOPIC("c9189", ERR, Logger::FIXME)
+        << "Unhandled exception while creating database";
+    return Result(TRI_ERROR_INTERNAL);
+  }
 
-    upgradeRes = methods::Upgrade::createDB(*vocbase, sanitizedUsers.slice());
+  if (ServerState::instance()->isCoordinator()) {
+    res = createCoordinator(createInfo);
+  } else {  // Single, DBServer, Agency
+    res = createOther(createInfo);
   }
 
-  if (upgradeRes.fail()) {
+  if (res.fail()) {
     LOG_TOPIC("1964a", ERR, Logger::FIXME)
-        << "Could not create database: " << upgradeRes.errorMessage();
-    return std::move(upgradeRes).result();
+        << "Could not create database: " << res.errorMessage();
+    return res;
   }
 
   // Entirely Foxx related:
+  // TODO: find out why this is here
```

Comment:
This is here because Foxx can only be executed on Coordinators and SingleServers not on DBServers
```suggestion

```